### PR TITLE
Fix inconsistent new install behaviour for templates when using block themes

### LINKED ISSUE #10530: Inconsistent New Install behaviour for templates when using block themes
Original report:

https://wordpress.org/support/topic/woocommerce-blocks-checkout-by-default-after-updating-to-8-0/

New installs sometimes have a block based checkout page rather than a shortcode based checkout page. It is not yet clear if this affects all installs, or is a race condition. 

Migration on init:
https://github.com/woocommerce/woocommerce-blocks/blob/5c99f331cf3a42339f7326eded47612b40e7ee3d/src/BlockTemplatesController.php#L82

Install on init:
https://github.com/woocommerce/woocommerce/blob/106e15f74a20e45cec9b92dd150c5c18c2c0f9a8/plugins/woocommerce/includes/class-wc-install.php#L254C42-L254C55

If this is confirmed to be a race condition, we can look into skipping migration when `WC_Installing` is true.

cc @wavvves @ralucaStan

### PR DESCRIPTION
**Pull request by wavvves**
This PR adds checks to postpone Cart and Checkout migrations, and also fixes a bug where migration would run, save content, but not attach it as a change to the template. The reason behind this was missing term creation , and therefore failing to set post terms.

Fixes #10530

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR has either a `[type]` label or a `[skip-changelog]` label.
- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Get a new env ready to install WP/WC using WP-CLI. You can use this [quick docker setup](https://github.com/woocommerce/woocommerce-blocks/files/12394720/new-test-env.zip)
2. Alias or copy  this branch, built ready to use as a plugin into the env
3. Using WP-CLI set up WC using the following command sequence (alter args to suit your case)

````
wp config create --dbname=wordpress --dbuser=wordpress --dbpass=wordpress --dbhost=db  --force 
wp core install --url=http://localhost --title='WP Staging' --admin_user=admin --admin_password=pass --admin_email=admin@wp.loc 
wp option set blog_public 0 
wp plugin activate woocommerce-blocks 
wp plugin install woocommerce --version=6.0.0 
wp plugin activate woocommerce 
wp plugin install wordpress-importer --activate 
wp import ./wp-content/plugins/woocommerce/sample-data/sample_products.xml --authors=skip 
wp plugin update woocommerce
````
3. Visit the store for the first time, head to Appearance > Editor > Templates > Manage all templates **verify that Cart and Checkout have customizations saved**
4. Open each one and verify they contain the shortcode version (matching page content)
5. Test the complete checkout flow.

* [ ] Do not include in the Testing Notes 
* [x] Should be tested by the development team exclusively 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Fixed a bug causing  Cart and Checkout Blocks to be used by default instead of shortcode versions on WC updates.

### CODE REVIEW

File: src/BlockTemplatesController.php [hunk 1]
```diff
@@ -764,8 +765,14 @@ function_exists( 'is_shop' ) &&
 
 	/**
 	 * Migrates page content to templates if needed.
+	 * This is skipped for requests regarding WP/WC setup and WP-ClI usage, to avoid issues with the pages not existing yet.
 	 */
 	public function maybe_migrate_content() {
+
+		if ( ! get_option( 'woocommerce_db_version', false ) || Constants::is_defined( 'WP_SETUP_CONFIG' ) || Constants::is_defined( 'WC_INSTALLING' ) || ( defined( 'WP_CLI' ) && WP_CLI ) ) {
```
Comment (mikejolley) [inline src/BlockTemplatesController.php]:
```suggestion
		if ( ! get_option( 'woocommerce_db_version', false ) || Constants::is_defined( 'WP_SETUP_CONFIG' ) || Constants::is_defined( 'WC_INSTALLING' ) || Constants::is_defined( 'WP_CLI' ) ) {
```

File: src/BlockTemplatesController.php [hunk 2]
```diff
@@ -767,6 +768,11 @@ function_exists( 'is_shop' ) &&
 	 * Migrates page content to templates if needed.
 	 */
 	public function maybe_migrate_content() {
+		// Postpone migration in these cases.
```
Comment (opr) [inline src/BlockTemplatesController.php]: Can we improve this comment to explain what we're checking and why migration should be skipped?

Someone just glancing over this code needs to take a moment to reason with each of the conditions, a written explanation reduces this effort required and would be super helpful.

**Review by opr — APPROVED** [review]
Works for me, and the code seems legit based on a discussion in slack.

To summarise our discussion: The changes to `create_custom_template` are necessary because without adding the theme slug to the `wp_theme` term, the customisations for this template will not be found when loading the template in `get_block_templates_from_db`.

Added one inline comment but it's easy to do so approving.

File: src/BlockTemplatesController.php (hunk 2 above)
Comment (wavvves) [reply src/BlockTemplatesController.php]: Done 👍🏼

**Pull request merged**